lib/internal/modules/esm/loader.js

```diff
@@ -202,33 +209,65 @@ class ESMLoader {
       const { ModuleWrap, callbackMap } = internalBinding('module_wrap');
       const module = new ModuleWrap(url, undefined, source, 0, 0);
       callbackMap.set(module, {
-        importModuleDynamically: (specifier, { url }) => {
-          return this.import(specifier, url);
+        importModuleDynamically: (specifier, { url }, import_assertions) => {
+          return this.import(specifier, url, import_assertions);
         }
       });
 
       return module;
     };
     const job = new ModuleJob(this, url, evalInstance, false, false);
-    this.moduleMap.set(url, job);
+    this.moduleMap.set(url, job, undefined);
     const { module } = await job.run();
 
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async getModuleJob(specifier, parentURL) {
+  async getModuleJob(specifier, parentURL, import_assertions) {
+    if (!ArrayPrototypeIncludes(supportedTypes, import_assertions.type)) {
+      throw new ERR_INVALID_IMPORT_ASSERTION('type', import_assertions.type);
+    }
     const { format, url } = await this.resolve(specifier, parentURL);
-    let job = this.moduleMap.get(url);
-    // CommonJS will set functions for lazy job evaluation.
-    if (typeof job === 'function') this.moduleMap.set(url, job = job());
 
-    if (job !== undefined) return job;
+    const jobMap = this.moduleMap.get(url);
+    let job;
+
+    if (jobMap != null) {
+      if (import_assertions.type != null && undefined in jobMap) {
+        job = jobMap[undefined];
+        // To avoid race conditions, always wait for non typed import to
```

Comment:
i think if you just got rid of the `Map<type, Module>`, you wouldn't have to have the weird dual logic here, and the one module would always have to either be untyped or match `import_assertions.type` here, which is what we want anyway.

Reply:
I'm not sure I understand. All concurrent jobs could fail or conflicting types may be requested:

```mjs
import('a', {assert: {type: "json"}});
import('a', {assert: {type: "javascript"}}); // bikeshed
```

Reply:
i'm saying our logic can just be `const job = getJob(); assert (assertion.type === undefined || assertion.type === job.type)`. the order in which imports occurs does not change what type the module will be.

Reply:
Done in https://github.com/nodejs/node/pull/39921/commits/1e6330a0fe881cb77c80a7b2166f89945a7daed8, PTAL.